torchdata/datapipes/iter/util/prefetcher.py

```diff
@@ -152,6 +150,17 @@ def resume(self):
             self.prefetch_data.run_prefetcher = True
             self.prefetch_data.paused = False
 
+    @final
+    def shutdown(self):
+        if self.prefetch_data:
+            self.prefetch_data.run_prefetcher = False
+            self.prefetch_data.stop_iteration = True
+        if self.thread:
+            self.thread.join()
```

Comment:
We need to make sure this function is called when new `iterator` is created. So, you might want to add the following lines to `__iter__`:
```py
if self.thread is not None:
    self.shutdown()
```

Reply:
I actually think it is handled by `reset` method, which is called whenever `__iter__` is called